File: src/main/java/net/csdn/trace/TraceContext.java
package net.csdn.trace;

import net.csdn.ServiceFramwork;
import net.csdn.annotation.FDesc;
import net.csdn.modules.log.SystemLogger;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import static net.csdn.common.collections.WowCollections.map;

/**
 * 3/29/14 WilliamZhu(allwefantasy@gmail.com)
 */
public class TraceContext {
    private String traceId;
    private String traceAppName;
    private String traceAPI;
    private String traceParent;
    private String traceSpan = "0";
    private AtomicInteger traceSpanInc = new AtomicInteger(-1);


    private long requestLen = 0;
    private long responseLen = 0;


    public TraceContext enterMethod(String methodName) {
        log("enter", methodName);
        return this;
    }


    public TraceContext exitMethod(String methodName) {
        log("exit", methodName);
        return this;
    }

    public static Map sendHttpTtraceContext(TraceContext context) {
        return map(
                "traceSpan", context.traceSpan + "." + context.traceSpanInc.incrementAndGet(),
                "traceId", context.traceId);

    }

    public static Trace receiveHttpTrateContext(Map items) {
        return null;
    }

    private void log(String enterOrExit, String methodName) {
        StringBuffer buffer = new StringBuffer();
        buffer.append(traceId + " " + traceAPI + " " + traceParent + " " + traceSpan + " " + traceAppName + " " + methodName);
        buffer.append(" " + enterOrExit);
        buffer.append(" " + System.currentTimeMillis());
        ServiceFramwork.injector.getInstance(SystemLogger.class).traceLogger().info(buffer.toString());
    }

    private TraceContext() {
    }

    public static TraceContext initial(
            @FDesc("系统名称，比如ServiceFramework") String type,
            @FDesc("服务/方法") String api,
            long requestLen,
            String traceId,
            String traceParent,
            String traceSpan
    ) {
        TraceContext context = new TraceContext();
        if (traceId == null) {
            context.traceId = generateTraceId();
            context.requestLen = requestLen;

        } else {
            context.traceId = traceId;
            context.traceParent = traceParent;
            //如果是其他系统的调用，那么该值会由请求系统递增过了的。
            context.traceSpan = traceSpan;
        }

        context.traceAppName = type;
        context.traceAPI = api;
        return context;
    }

    private static String generateTraceId() {
        return UUID.randomUUID().toString();
    }

}

File: src/main/java/net/csdn/trace/TraceContext.scala
package net.csdn.trace

import java.util.UUID

import net.csdn.common.logging.CSLogger
import net.csdn.trace.VisitType.VisitType

/**
 * 10/14/15 WilliamZhu(allwefantasy@gmail.com)
 */
object TraceContext {

  def createRemoteContext = {
    val traceId = UUID.randomUUID().getMostSignificantBits + ""
    val context = new TraceContext(traceId)
    context
  }

  def parseRemoteContext(params: java.util.Map[String, Array[String]]) = {
    val traceId = params.get(RemoteTraceElementKey.TRACEID)(0)
    val rpcId = params.get(RemoteTraceElementKey.RPCID)(0) + ".0"
    val context = new TraceContext(traceId)
    context.rpcId = rpcId
    context
  }

  def createRemoteContextForNewThread(traceId: String, rpcId: String) = {
    val context = new TraceContext(traceId)
    context.rpcId = rpcId
    context
  }


}

class TraceContext(val traceId: String) {

  var remoteTraceElement: RemoteTraceElement = _

  def configRemoteTraceElement(rte: RemoteTraceElement) = {
    remoteTraceElement = rte
  }

  def newRemoteTraceElement(door: Boolean, rpcId: String, url: String, visitType: VisitType) = {
    RemoteTraceElement(TraceId(traceId, door, 0), rpcId, url, System.nanoTime(), 0, 0, 0, "", visitType)
  }

  def start(url: String, visitType: VisitType) = {
    remoteTraceElement = newRemoteTraceElement(false, rpcId, url, visitType)
  }

  def openDoor(rpcId: String, url: String, visitType: VisitType) = {
    remoteTraceElement = newRemoteTraceElement(true, rpcId, url, visitType)
  }

  def finish(_resultCode: Int, responseLength: Long, _message: String, logger: CSLogger) = {
    val _startTime = remoteTraceElement.startTime
    val _oldMessage = remoteTraceElement.message
    val rte = remoteTraceElement.copy(
      traceId = TraceId(traceId, false, 0),
      timeConsume = (System.nanoTime() - _startTime),
      resultCode = _resultCode,
      message = if (_message != null) _message else _oldMessage,
      length = responseLength
    )
    remoteTraceElement = null
    log(logger, rte)
  }

  def log(logger: CSLogger, rte: RemoteTraceElement) = {
    logger.info("___trace___\t" + s"${rte.traceId.id}\t${rte.traceId.door}\t${rte.visitType.toString}\t${rte.rpcId}\t${rte.url}\t${rte.startTime}\t${rte.timeConsume}\t${rte.resultCode}\t${rte.length}\t${rte.message}")
  }

  var rpcId = "0.0"

  def currentRpcId = rpcId

  def nextRpcId = {
    synchronized {
      val phase = rpcId.split("\\.")
      rpcId = rpcId.take(phase.length - 1) + "." + (phase.last.toInt + 1)
      rpcId
    }
  }
}
